src/main/java/org/opensearch/security/http/OnBehalfOfAuthenticator.java

```diff
@@ -0,0 +1,249 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.OpenSearchException;
+import org.opensearch.OpenSearchSecurityException;
+import org.opensearch.SpecialPermission;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.rest.RestChannel;
+import org.opensearch.rest.RestRequest;
+import org.opensearch.security.auth.HTTPAuthenticator;
+import org.opensearch.security.authtoken.jwt.EncryptionDecryptionUtil;
+import org.opensearch.security.ssl.util.ExceptionUtils;
+import org.opensearch.security.user.AuthCredentials;
+import org.opensearch.security.util.KeyUtils;
+
+import static org.opensearch.security.OpenSearchSecurityPlugin.LEGACY_OPENDISTRO_PREFIX;
+import static org.opensearch.security.OpenSearchSecurityPlugin.PLUGINS_PREFIX;
+
+public class OnBehalfOfAuthenticator implements HTTPAuthenticator {
+
+    private static final String REGEX_PATH_PREFIX = "/(" + LEGACY_OPENDISTRO_PREFIX + "|" + PLUGINS_PREFIX + ")/" + "(.*)";
+    private static final Pattern PATTERN_PATH_PREFIX = Pattern.compile(REGEX_PATH_PREFIX);
+    private static final String ON_BEHALF_OF_SUFFIX = "api/generateobotoken";
+    private static final String ACCOUNT_SUFFIX = "api/account";
+
+    protected final Logger log = LogManager.getLogger(this.getClass());
+
+    private static final Pattern BEARER = Pattern.compile("^\\s*Bearer\\s.*", Pattern.CASE_INSENSITIVE);
+    private static final String BEARER_PREFIX = "bearer ";
+
+    private final JwtParser jwtParser;
+    private final String encryptionKey;
+    private final Boolean oboEnabled;
+    private final String clusterName;
+
+    public OnBehalfOfAuthenticator(Settings settings, String clusterName) {
+        String oboEnabledSetting = settings.get("enabled");
+        oboEnabled = oboEnabledSetting == null ? Boolean.TRUE : Boolean.valueOf(oboEnabledSetting);
+        encryptionKey = settings.get("encryption_key");
+        jwtParser = initParser(settings.get("signing_key"));
+        this.clusterName = clusterName;
+    }
+
+    private JwtParser initParser(final String signingKey) {
+        JwtParser _jwtParser = KeyUtils.createJwtParserFromSigningKey(signingKey, log);
+
+        if (_jwtParser == null) {
+            throw new RuntimeException("Unable to find on behalf of authenticator signing key");
+        }
+
+        return _jwtParser;
+    }
+
+    private List<String> extractSecurityRolesFromClaims(Claims claims) {
+        Object er = claims.get("er");
+        Object dr = claims.get("dr");
+        String rolesClaim = "";
+
+        if (er != null) {
+            rolesClaim = EncryptionDecryptionUtil.decrypt(encryptionKey, er.toString());
+        } else if (dr != null) {
+            rolesClaim = dr.toString();
+        } else {
+            log.warn("This is a malformed On-behalf-of Token");
+        }
+
+        List<String> roles = Arrays.stream(rolesClaim.split(","))
```

Comment:
Can roles names have a comma in them? That could allow for strange interpretations, this might be a good follow up issue that should be resolved before release.

Reply:
Based on:
> The role name must contain from 2 to 50 characters. Valid characters are A-Z, a-z, 0-9, (_)underscore, (-) hyphen and unicode characters.

I don't think the role names should contain comma in them.